src/sync.rs

```diff
@@ -460,40 +460,47 @@ pub fn sync(
             }
 
             local.end_atomic().context(EndAtomicSnafu {})?;
-            Ok(())
+            Ok(new_local_emails)
         };
 
-        if let Err(e) = commit_changes() {
-            // Remove all the symlinks.
-            for new_email in &new_emails {
-                debug!(
-                    "Removing symlink `{}'",
-                    &new_email.maildir_path.to_string_lossy(),
-                );
-                if let Err(e) = fs::remove_file(&new_email.maildir_path) {
-                    warn!(
-                        "Could not remove symlink `{}': {e}",
+        let new_local_emails = match commit_changes() {
+            Ok(r) => r,
+            Err(e) => {
+                // Remove all the symlinks.
+                for new_email in &new_emails {
+                    debug!(
+                        "Removing symlink `{}'",
                         &new_email.maildir_path.to_string_lossy(),
                     );
+                    if let Err(e) = fs::remove_file(&new_email.maildir_path) {
+                        warn!(
+                            "Could not remove symlink `{}': {e}",
+                            &new_email.maildir_path.to_string_lossy(),
+                        );
+                    }
                 }
+                // Fail as normal.
+                return Err(e);
             }
-            // Fail as normal.
-            return Err(e);
-        }
+        };
 
         // Now that the atomic database operation has been completed, do the actual file operations.
 
         // Replace the symlinks with the real files.
         for new_email in &new_emails {
+            // We have to go back to the notmuch `Message' object to get the path, as it may have
+            // modified the path away from our original to include the flags.
+            let local_email = new_local_emails.get(&new_email.remote_email.id).unwrap();
+            let new_maildir_path = local_email.message.filename();
```

Comment:
This has the potential for an edge case if there are duplicate messages which are indexed as the same message by notmuch. (For example, if you send a message to a mailing list and that list sends you back a copy of your own message.) There's some code in local.rs that specifically matches filenames via a regex for a maildir path which handles similar cases.

Reply:
This is interesting.

I've tried to create a test case for that by manually inserting two messages into the store that have the same message ID, then running mujmap. They both went in just fine; `notmuch search id:xxx` shows them both, they're accessible if you fiddle with `--duplicates`, and the paths are all hooked up right. Later runs updates only one of them, of course, but that's the same behaviour as before I think.

So .. I'm not sure. Is there another kind of "duplicate" that _isn't_ based on the id in notmuch?

Reply:
I pushed some commits to main last night which fixed a similar issue. Imagine two JMAP messages with unique JMAP IDs but the same notmuch ID. When copying the mail files to replace the symlinks here, `filename()` will only consider a single arbitrary filename, which will cause two copies to be performed to the same filename and for the second symlink to be ignored. Worse, this could change the contents of that file unexpectedly, if the duplicate messages have different blob IDs, which they likely do.

Reply:
Right, I've read all the way down into notmuch proper and I think I finally get it.

I thought that duplicate ids would give two separate notmuch message record, each with their own filenames, that were just masked by the fact that you couldn't reference (search for) them. What's actually happening is that there's just one notmuch record, maybe with multiple filenames.

And that's why I'm wrong in thinking the `Message` we get back from `index_file` would obviously give the right name for `filename()`, when it was just giving one of them. Presumably, most recently seen, which is why it was working for me.

Its not that I didn't believe you, I just wanted to understand why :D Thanks!

Reply:
Ok, think I got it!